In [84]:
import pandas as pd 
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

In [85]:
df = pd.read_csv('../data/Titanic_processed.csv') #Stored the processed data in df


In [86]:
features = df.drop(columns=['Survived'])
labels = df['Survived']

In [87]:
X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=0.3, random_state=42)

In [88]:
# -------------------------------------------------------------------------- #

# model = LogisticRegression(max_iter= 1000) #Initializing the model
# model.fit(X_train, y_train) #Applying the fit method
# predictions = model.predict(X_test)

# -------------------------------------------------------------------------- #

# Implementing Standard Scaler to improve accuracy

from sklearn.preprocessing import StandardScaler

scalar = StandardScaler()

X_train_scaled = scalar.fit_transform(X_train)
X_test_scaled = scalar.transform(X_test)

model = LogisticRegression()
model.fit(X_train_scaled, y_train)
predictions = model.predict(X_test_scaled)

In [89]:
from sklearn.metrics import confusion_matrix

c_matrix = confusion_matrix(y_test, predictions)
print(c_matrix, end = '\n\n')

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, predictions)
print(f"Accuracy: {accuracy * 100:.2f}%", end = '\n\n')

print(predictions)

[[138  29]
 [ 25  75]]

Accuracy: 79.78%

[0 1 1 0 1 0 0 0 1 1 0 1 0 0 0 0 0 0 0 0 0 1 0 1 0 1 0 1 0 0 1 1 0 1 0 0 1
 1 0 0 0 1 0 0 0 1 1 0 0 1 1 1 0 0 1 1 1 0 0 0 0 1 1 0 1 0 0 0 1 1 0 1 1 0
 0 1 0 0 1 1 0 1 1 1 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 1 1 0 0 0 1 1 0 0 0
 1 0 1 0 0 0 0 1 0 0 1 0 0 1 1 1 1 1 0 1 0 0 0 1 1 0 1 1 1 1 0 1 0 1 0 0 1
 0 1 0 1 0 1 0 1 1 0 0 1 0 1 0 0 0 1 1 0 0 1 1 1 1 0 1 0 0 1 1 1 0 0 0 1 0
 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 1 1 0 0 0 1 0 0 1 0 0 1 0 1
 1 1 1 0 0 1 1 1 1 0 0 0 0 0 0 1 0 0 1 0 1 0 0 0 1 0 0 0 0 0 0 1 0 0 0 1 0
 0 0 0 0 1 0 0 1]


In [90]:
from sklearn.metrics import classification_report

print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       0.85      0.83      0.84       167
           1       0.72      0.75      0.74       100

    accuracy                           0.80       267
   macro avg       0.78      0.79      0.79       267
weighted avg       0.80      0.80      0.80       267



## Project Conclusion & Model Evaluation 📊

After cleaning the data, engineering features, and applying standard scaling, the Logistic Regression model achieved a baseline **Accuracy of ~80%**.

### Key Insights from the Classification Report:
* **Identifying Non-Survivors (Class 0):** The model is highly confident and accurate when predicting passengers who did not survive,showing 
**85% Precision** and **83% Recall**.
* **Identifying Survivors (Class 1):** Performance dips slightly when predicting survivors (**72% Precision**, **75% Recall**). 
* **The "Class Imbalance" Effect:** This discrepancy is a natural result of the historical data. Because a larger percentage of passengers unfortunately perished, the algorithm had significantly more training examples to learn the mathematical profile of a "non-survivor."

### Next Steps & Future Improvements
While 80% is a strong and mathematically sound baseline for Logistic Regression on this dataset, future iterations of this analysis could include:
1. **Advanced Feature Engineering:** Extracting titles (e.g., *Dr., Rev., Mrs.*) from the dropped `Name` column to capture social status and priority boarding rules.
2. **Algorithm Exploration:** Testing non-linear algorithms, such as **Random Forest Classifiers** or **Support Vector Machines (SVM)**, which do not rely on drawing straight mathematical lines and might capture more complex relationships within the passenger data.